## Importing Required Packages and Dependiences

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
#Now lets load credit cada data using pandas
cc_data=pd.read_csv("creditcard.csv")

### Content about dataset

The dataset contains transactions made by credit cards in September 2013 by European cardholders.
This dataset presents transactions that occurred in two days, where we have 492 frauds out of 284,807 transactions. The dataset is highly unbalanced, the positive class (frauds) account for 0.172% of all transactions.

It contains only numerical input variables which are the result of a PCA transformation. Unfortunately, due to confidentiality issues, we cannot provide the original features and more background information about the data. Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value 1 in case of fraud and 0 otherwise.

In [3]:
#dataset info
cc_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [4]:
#head of dataset
cc_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [5]:
#data set tail
cc_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [6]:
#Lets see if we have missing values in the data
cc_data.isnull().sum()  #count of missing values in each column

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

No Missing Values. So it is very good thing, if we had missing values we should deal with it

In [7]:
#Lets check the distribution of class
cc_data["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

This dataset is highly unbalanced as we have 492 frauds and 284315 normal ones

## Seperating Data for Analysis 

In [8]:
normal=cc_data[cc_data.Class==0] #all normal transactions into one
fraud=cc_data[cc_data.Class==1] #all fraud transactions into one

### Lets look at some statistics 

In [9]:
#Summary(describe) of normal Amount
normal.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [10]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [11]:
#lets compare all transactions by class 
cc_data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

### Lets build a balanced dataset 

As we have 492 rows in fraud, lets take 492 transactions from normal to get a balanced dataset

In [12]:
normal_sample = normal.sample(n=492)  #takes randomly 492

In [13]:
new_ds = pd.concat([normal_sample,fraud],axis=0) #axis=0 add row wise , axis=1 add column wise

In [14]:
new_ds.head()

Time        V1        V2        V3        V4        V5        V6  \
234683  148069.0  1.879081 -0.247422 -0.789010  1.449703  0.058621  0.349005   
108267   70831.0  1.116428  0.142792  0.311588  1.083399 -0.203527 -0.253299   
66622    52136.0  0.867381 -0.494472  0.966811  1.320533 -0.871792  0.299731   
75201    55949.0  1.256500  0.412545  0.322783  0.685814 -0.299753 -1.054033   
94089    64726.0  1.400351 -1.410898  0.267343 -1.642181 -1.324857  0.196671   

              V7        V8        V9  ...       V21       V22       V23  \
234683 -0.193549  0.159021  0.891548  ... -0.278284 -0.522928  0.256883   
108267 -0.003529  0.203112 -0.060817  ... -0.414169 -1.403603  0.260930   
66622  -0.476007  0.197105  0.463161  ...  0.241196  0.507572 -0.269776   
75201   0.142000 -0.249935 -0.090218  ... -0.278160 -0.763634  0.119480   
94089  -1.312387  0.083147 -1.724764  ... -0.046957  0.070010 -0.215268   

             V24       V25       V26       V27       V28  Amount  Class  
234683  0.662547  0.011546 -0.870127  0.028301 -0.044493   23.95      0  
108267 -0.137152  0.057053 -0.816990  0.004482  0.012940   19.49      0  
66622   0.021172  0.466840 -0.254398  0.036272  0.047524  150.32      0  
75201   0.367562  0.243414  0.091667 -0.017813  0.032599    1.98      0  
94089  -0.852429  0.426617 -0.054488  0.030667  0.016748   83.20      0  

[5 rows x 31 columns]

In [15]:
new_ds.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [16]:
new_ds["Class"].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [17]:
new_ds.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92202.239837 -0.147048  0.067125  0.092958  0.044290  0.049975   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.123460 -0.070760 -0.060075 -0.026621  ... -0.068831  0.012297   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.033798  0.067670  0.034419  0.010831  0.024176  0.039471  0.013056   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       62.750203  
1      122.211321  

[2 rows x 30 columns]

We can observe we dont have large changes in this groupby comparison vs last groupby comparison

### Splitting Data into Features and Targets 

In [18]:
X=new_ds.drop(columns="Class",axis=1) #drop Class column
Y=new_ds["Class"]

In [19]:
X.head()

Time        V1        V2        V3        V4        V5        V6  \
234683  148069.0  1.879081 -0.247422 -0.789010  1.449703  0.058621  0.349005   
108267   70831.0  1.116428  0.142792  0.311588  1.083399 -0.203527 -0.253299   
66622    52136.0  0.867381 -0.494472  0.966811  1.320533 -0.871792  0.299731   
75201    55949.0  1.256500  0.412545  0.322783  0.685814 -0.299753 -1.054033   
94089    64726.0  1.400351 -1.410898  0.267343 -1.642181 -1.324857  0.196671   

              V7        V8        V9  ...       V20       V21       V22  \
234683 -0.193549  0.159021  0.891548  ... -0.262026 -0.278284 -0.522928   
108267 -0.003529  0.203112 -0.060817  ... -0.299289 -0.414169 -1.403603   
66622  -0.476007  0.197105  0.463161  ...  0.181958  0.241196  0.507572   
75201   0.142000 -0.249935 -0.090218  ... -0.028457 -0.278160 -0.763634   
94089  -1.312387  0.083147 -1.724764  ... -0.151414 -0.046957  0.070010   

             V23       V24       V25       V26       V27       V28  Amount  
234683  0.256883  0.662547  0.011546 -0.870127  0.028301 -0.044493   23.95  
108267  0.260930 -0.137152  0.057053 -0.816990  0.004482  0.012940   19.49  
66622  -0.269776  0.021172  0.466840 -0.254398  0.036272  0.047524  150.32  
75201   0.119480  0.367562  0.243414  0.091667 -0.017813  0.032599    1.98  
94089  -0.215268 -0.852429  0.426617 -0.054488  0.030667  0.016748   83.20  

[5 rows x 30 columns]

In [20]:
print(Y)

234683    0
108267    0
66622     0
75201     0
94089     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


## Split into training and testing data 

In [21]:
#using the function we imported above
#split is random
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=30)

In [22]:
print(X.shape,X_train.shape,X_test.shape) #984 are divided into 787 and 197

(984, 30) (787, 30) (197, 30)


# Model Training Using Logistic regression

We can also use different models but as it is binary classification we used logistic regression 

In [23]:
model = LogisticRegression(solver='lbfgs', max_iter=1000) #by default max_iter=100

In [24]:
#training the model with training data
model.fit(X_train,Y_train)

LogisticRegression(max_iter=1000)

## Model evaluation

In [25]:
#accuracy on training data
X_train_prediction = model.predict(X_train)
td_accuracy = accuracy_score(X_train_prediction,Y_train)

In [26]:
print("Accuracy Of Training Data : ",td_accuracy)

Accuracy Of Training Data :  0.9072426937738246


In [27]:
test_accuracy = accuracy_score(model.predict(X_test),Y_test)

In [28]:
print("Accuracy Of Testing Data : ",test_accuracy)

Accuracy Of Testing Data :  0.9390862944162437
